plugins/inputs/solr/solr.go

```diff
@@ -5,378 +5,165 @@ import (
-		"99th_pc_request_time":       getFloat(value["99thPcRequestTime"]),
-		"999th_pc_request_time":      getFloat(value["999thPcRequestTime"]),
-		"avg_requests_per_second":    getFloat(value["avgRequestsPerSecond"]),
-		"avg_time_per_request":       getFloat(value["avgTimePerRequest"]),
-		"errors":                     getInt(value["errors"]),
-		"handler_start":              getInt(value["handlerStart"]),
-		"median_request_time":        getFloat(value["medianRequestTime"]),
-		"requests":                   getInt(value["requests"]),
-		"timeouts":                   getInt(value["timeouts"]),
-		"total_time":                 getFloat(value["totalTime"]),
-	}
-}
-
-// Add update metrics section to accumulator
-func addUpdateHandlerMetricsToAcc(acc telegraf.Accumulator, core string, mBeansData *MBeansData, measurementTime time.Time) error {
-	var updateMetrics map[string]UpdateHandler
-
-	if len(mBeansData.SolrMbeans) < 6 {
-		return fmt.Errorf("no update handler metric data to unmarshal")
-	}
-	if err := json.Unmarshal(mBeansData.SolrMbeans[5], &updateMetrics); err != nil {
-		return err
-	}
-	for name, metrics := range updateMetrics {
-		var autoCommitMaxTime int64
-		if len(metrics.Stats.AutocommitMaxTime) > 2 {
-			autoCommitMaxTime, _ = strconv.ParseInt(metrics.Stats.AutocommitMaxTime[:len(metrics.Stats.AutocommitMaxTime)-2], 0, 64)
-		}
-		coreFields := map[string]interface{}{
-			"adds":                        metrics.Stats.Adds,
-			"autocommit_max_docs":         metrics.Stats.AutocommitMaxDocs,
-			"autocommit_max_time":         autoCommitMaxTime,
-			"autocommits":                 metrics.Stats.Autocommits,
-			"commits":                     metrics.Stats.Commits,
-			"cumulative_adds":             metrics.Stats.CumulativeAdds,
-			"cumulative_deletes_by_id":    metrics.Stats.CumulativeDeletesByID,
-			"cumulative_deletes_by_query": metrics.Stats.CumulativeDeletesByQuery,
-			"cumulative_errors":           metrics.Stats.CumulativeErrors,
-			"deletes_by_id":               metrics.Stats.DeletesByID,
-			"deletes_by_query":            metrics.Stats.DeletesByQuery,
-			"docs_pending":                metrics.Stats.DocsPending,
-			"errors":                      metrics.Stats.Errors,
-			"expunge_deletes":             metrics.Stats.ExpungeDeletes,
-			"optimizes":                   metrics.Stats.Optimizes,
-			"rollbacks":                   metrics.Stats.Rollbacks,
-			"soft_autocommits":            metrics.Stats.SoftAutocommits,
+	version, err := s.determineServerAPIVersion(server)
+	if err != nil {
+		s.Log.Errorf("Getting version for %q failed: %v", server, err)
+		// Exit early and do not fill the cache as the server might not be
+		// reachable yet.
+		return newAPIv1Config()
+	}
+	s.Log.Debugf("Found API version %d for server %q...", version, server)
+
+	switch version {
+	case 0:
+		s.Log.Warn("Unable to determine API version! Using API v1...")
+		fallthrough
+	case 1:
+		s.configs[server] = newAPIv1Config()
+	case 2:
+		fallthrough
```

Comment:
From a maintenance/style standpoint I would prefer we do what we need to do in case 0 and 2 and break rather than fallthrough. It is too easy for someone to come by at a later date and add in another case here for version 3 and/or modify the default method not realizing it is dependent on case 2. I realize the trade off is some duplicated calls.

This is minor and a personal nit so up to you if you want to change ;)